File: Java-Kotlin/src/main/kotlin/twopointers/Longest Substring Without Repeating Characters.kt
package twopointers

import java.util.HashSet

/**
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * */

fun lengthOfLongestSubstring(s: String): Int {
    var ans = 0
    var left = 0
    val set = HashSet<Char>()
    for (right in s.indices) {
        while (set.contains(s[right]))
            set.remove(s[left++])
        ans = Math.max(ans, right - left + 1)
        set.add(s[right])
    }
    return ans
}

File: C++/Sliding Window/Longest Substring Without Repeating Characters.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * */

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> pos(255, -1);
        int result = 0, left = 0;
        for (int right = 0; right < s.length(); right++) {
            char chr = s[right];
            if (pos[chr] != -1 && pos[chr] >= left)
                left = pos[chr] + 1;
            result = max(result, right - left + 1);
            pos[chr] = right;
        }
        return result;
    }
};
